flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/RowSerializer.java

```diff
@@ -367,21 +367,21 @@ public int getVersion() {
 	/**
 	 * A {@link TypeSerializerSnapshot} for RowSerializer.
 	 */
-	// TODO not fully functional yet due to FLINK-17520
 	public static final class RowSerializerSnapshot extends CompositeTypeSerializerSnapshot<Row, RowSerializer> {
 
 		private static final int VERSION = 3;
 
-		private static final int VERSION_WITHOUT_ROW_KIND = 2;
+		private static final int LAST_VERSION_WITHOUT_ROW_KIND = 2;
 
-		private boolean legacyModeEnabled = false;
+		private int readVersion = VERSION;
 
 		public RowSerializerSnapshot() {
 			super(RowSerializer.class);
 		}
 
 		RowSerializerSnapshot(RowSerializer serializerInstance) {
 			super(serializerInstance);
+			this.readVersion = serializerInstance.legacyModeEnabled ? LAST_VERSION_WITHOUT_ROW_KIND : VERSION;
```

Comment:
I don't think this line is needed, unless I'm missing something in the tests.
The read version should only ever be changed if this snapshot was created by restoring from a snapshot.
In this case, this constructor is only ever used to create a new snapshot when checkpointing occurs - the read version should be the default value (`VALUE`).